prediction_market_agent/agents/goal_manager.py

```diff
@@ -0,0 +1,295 @@
+            limit=limit
+        )
+        return [EvaluatedGoal.from_model(model) for model in evaluated_goal_models]
+
+    def generate_goal(self, latest_evaluated_goals: list[EvaluatedGoal]) -> Goal:
+        """
+        Generate a new goal based on the high-level description and the latest
+        evaluated goals.
+
+        TODO support generation of long-horizon goals with a specified
+        completion date, until which the goal's status is 'pending'.
+        """
+        parser = PydanticOutputParser(pydantic_object=Goal)
+        prompt = PromptTemplate(
+            template=GENERATE_GOAL_PROMPT_TEMPLATE,
+            input_variables=[
+                "high_level_description",
+                "agent_capabilities",
+                "previous_evaluated_goals",
+            ],
+            partial_variables={"format_instructions": parser.get_format_instructions()},
+        )
+        latest_evaluated_goals_str = self.evaluated_goals_to_str(latest_evaluated_goals)
+        llm = ChatOpenAI(
+            temperature=0,
+            model=self.model,
+            api_key=APIKeys().openai_api_key_secretstr_v1,
+        )
+        chain = prompt | llm | parser
+
+        goal: Goal = chain.invoke(
+            {
+                "high_level_description": self.high_level_description,
+                "agent_capabilities": self.agent_capabilities,
+                "previous_evaluated_goals": latest_evaluated_goals_str,
+            }
+        )
+        return goal
+
+    def have_reached_retry_limit(
+        self, latest_evaluated_goals: list[EvaluatedGoal]
+    ) -> bool:
+        if self.retry_limit == 0:
+            return True
+
+        if len(latest_evaluated_goals) < self.retry_limit + 1:
+            return False
+
+        latest_goal = latest_evaluated_goals[0].to_goal()
+        return all(
+            g.to_goal() == latest_goal
+            for g in latest_evaluated_goals[: self.retry_limit + 1]
+        )
+
+    def get_goal(self) -> Goal:
+        """
+        Manage the fetching of goals from memory, and deciding when to generate
+        a new goal vs. retrying an incomplete one.
+
+        TODO add the ability to continue from a previous session if the goal
+        is not complete.
+        """
+        latest_evaluated_goals = self.get_latest_evaluated_goals_from_memory(
+            limit=self.retry_limit
+        )
+        if latest_evaluated_goals:
+            evaluated_goal = latest_evaluated_goals[0]
+            if evaluated_goal.is_complete:
+                # Generate a new goal
+                return self.generate_goal(latest_evaluated_goals)
+            else:
+                # Try again, unless we've reached the retry limit
+                if self.have_reached_retry_limit(latest_evaluated_goals):
+                    return self.generate_goal(latest_evaluated_goals)
+                else:
+                    return evaluated_goal.to_goal()
```

Comment:
Sorry, I'm a bit confused - how are you retrying? What constitutes an attempt - one iteration of the agent?
I don't love all these nested if-else, so wondering if we can simplify this.

Reply:
Clarified during stand-up -> retrying is checked/done after general agent has completed N iterations, after which goal is saved in the DB.
So flow is
-> create goal (contains logic for `retry` and `fetching same goal if it wasn't yet completed`)
-> do N iterations
-> store goal
-> create goal (same as above)
-> do N iterations
...
@evangriffiths can you confirm my understanding here?
If correct, please close this thread.

Reply:
Yes that's right! Sorry for leaving you hanging 🙏